cmd/skaffold/app/cmd/diagnose.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2018 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/version"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/watch"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdDiagnose describes the CLI command to diagnose skaffold.
+func NewCmdDiagnose(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "diagnose",
+		Short: "Run a diagnostic on Skaffold",
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return doDiagnose(out)
+		},
+	}
+	cmd.Flags().StringVarP(&opts.ConfigurationFile, "filename", "f", "skaffold.yaml", "Filename or URL to the pipeline file")
+	return cmd
+}
+
+func doDiagnose(out io.Writer) error {
+	_, config, err := newRunner(opts)
+	if err != nil {
+		return errors.Wrap(err, "creating runner")
+	}
+
+	fmt.Fprintln(out, "Skaffold version:", version.Get().GitCommit)
+	fmt.Fprintln(out, "Configuration version:", config.APIVersion)
+	fmt.Fprintln(out, "Number of artifacts:", len(config.Build.Artifacts))
+
+	fmt.Fprintln(out, "\nRunning a diagnostic on artifacts and their dependencies:")
+
+	if err := diagnoseArtifacts(out, config.Build.Artifacts); err != nil {
+		return errors.Wrap(err, "running diagnostic on artifacts")
+	}
+
+	fmt.Fprintln(out, "\nSecond run (that benefits from cache):")
```

Comment:
an idea: what if we list the result of the runs for each metric, instead of printing a full run? 
i.e.: 

````
Number of artifacts: 2

Running a diagnostic on artifacts and their dependencies:
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 4.809915ms
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 19.012µs
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 454.866µs
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 16.534µs

Second run (that benefits from cache):
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 396.262µs
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 16.35µs
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 305.815µs
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 29.689µs
````

it would be this: 

````
Number of artifacts: 2

Running a diagnostic on artifacts and their dependencies:
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 4.809915ms, 396.262µs (cache)
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 19.012µs,   16.35µs (cache)
Type: Docker
 - Dependencies to watch: 2
 - Time to list dependencies: 454.866µs, 305.815µs (cached)
 - Size of the Docker context: 3072
 - Time to compute mTimes for all dependencies: 16.534µs, 29.689µs (cached)

````

Reply:
I'll +1 this, the output of this can get pretty large for bigger projects